src/Castle.Facilities.AspNet.WebApi/AspNetWebApiFacility.cs

```diff
@@ -0,0 +1,108 @@
+// Copyright 2004-2017 Castle Project - http://www.castleproject.org/
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Castle.Facilities.AspNet.WebApi
+{
+	using System;
+	using System.Web.Http;
+	using System.Web.Http.Dispatcher;
+
+	using Castle.MicroKernel;
+	using Castle.MicroKernel.Facilities;
+
+	public class AspNetWebApiFacility : AbstractFacility, ISupportFacility
+	{
+		private bool isSelfHosted;
+		private bool initialised = false;
+		private HttpConfiguration httpConfiguration = null;
+		private Action<IKernel, Type> beforeControllerResolved;
+		private Action<IKernel, object> afterControllerReleased;
+		private AspNetWebApiControllerActivator controllerActivator;
+		private AspNetWebApiDependencyResolver dependencyResolver;
+
+		public new IKernel Kernel => base.Kernel;
+		public bool AutoCreateLifestyleScopes { get; private set; }
+
+		public AspNetWebApiControllerActivator ControllerActivator => controllerActivator;
+		public AspNetWebApiDependencyResolver DependencyResolver => dependencyResolver;
+
+		protected override void Init()
+		{
+			if (!isSelfHosted)
+			{
+				controllerActivator = new AspNetWebApiControllerActivator(this);
```

Comment:
I hate to keep making comparisons against Autofac, but I know nothing about ASP.NET Web API as I've spent all my time with Nancy, however the Autofac integration does both IIS and self-hosting with an `IDependencyResolver`, what am I missing?:

http://autofac.readthedocs.io/en/latest/integration/webapi.html#get-the-httpconfiguration

Reply:
I have had varied results getting this approach to work over the years under web hosts. In theory it should be fine but there have been cases where things changed over time and this simply was not reliable enough. I also wanted a separate extension for the webhost scope implementation.

Reply:
I could optimise this. Let me check it out again.